### ISSUE TITLE
Implement function overloading for user functions in the shader language

### ISSUE THREAD

**Issue by PureAsbestos** [godotengine/godot-proposals#2808]
### Describe the project you are working on
A raymarcher.

### Describe the problem or limitation you are having in your project
If I want to be able to call a custom function with different data types in a shader, like so:
```glsl
vec2 result_vec2 = some_func(some_vec2);
vec3 result_vec3 = some_func(some_vec3);
```
I can't, and must resort to giving the functions different names.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation
The solution is to add function overloading to Godot's shader language, just like in GLSL.
This is the same as this issue on the main repo from a few years ago: https://github.com/godotengine/godot/issues/10474
Now instead of failing silently, you just get an error message (`Redefinition of 'some_func'`).

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams
Allow function overloading. You could define two functions with the same name, but different arguments:
```glsl
vec2 some_func(vec2 some_vec2) {  /* some code here */ }
vec3 some_func(vec3 some_vec3) {  /* some code here */ }
```

### If this enhancement will not be used often, can it be worked around with a few lines of script?
It will be used fairly often and can't be easily worked around.

### Is there a reason why this should be core and not an add-on in the asset library?
Probably not possible with an add-on. Provides better feature parity with GLSL and more orthogonality in general since built-ins already use overloading.

**Comment by briansemrau** [comment]
This is something that is quite painful for me to not have. I have a collection of noise and utility functions I like to use with many `vecn` type variants. I'm often extending and modifying it so I prefer to keep it as clean glsl to be used wherever. But can't just paste them in to a gdshader; I have to refactor hundreds of lines of code.

**Comment by hmans** [comment]
I was surprised by this today when porting some GLSL library code over to Godot; I'm actually surprised that this hasn't received more upvotes (or seen more activity.) Eep!

Yes, very much in favor of this proposal. 👍

**Comment by Calinou** [comment]
@hmans Please don't bump issues without contributing significant new information. Use the :+1: reaction button on the first post instead.

**Comment by MoltenCoffee** [comment]
Out of curiosity, is this an issue that just hasn't seen enough time by a contributor dedicated to it, or are there any major roadblocks that prevent this from being implemented at the moment?

**Comment by AThousandShips** [comment]
No one has picked it up, either no one is interested, or no one has solved it, or no one has time

**Comment by hmans** [comment]
> Out of curiosity, is this an issue that just hasn't seen enough time by a contributor dedicated to it, or are there any major roadblocks that prevent this from being implemented at the moment?

At this point, I simply don't think enough people are actually authoring Godot shaders in any significant fashion, or else someone would have implemented a fix by now (also see #74817, which includes some absurd suggestions for workarounds by a certain Godot maintainer who seems to have an issue with me in particular.)

Like my previous comment, this one will probably be hidden, but I am beyond caring. There used to be a time where I still had hope for this engine, but it has long faded.
